File: python ver/status_effect.py
"""
Status effects include Weaken (lowers attack), Decay (lowers defense), Slow (lowers speed), Disquiet (lowers magic), 
Poison (lowers HP per turn) and Freeze (skips a turn)
Positive status effects do the opposite of this.
Status effects heal after a certain amount of turns.
"""
from enum import Enum
from random import *

class StatusEffect(Enum):
    # actual effects should be handled in entity class
    # negative
    WEAKEN = "Weaken" # lower attack
    DECAY = "Decay" # lower defense
    SLOW = "Slow" # lower speed
    DISQUIET = "Disquiet" # lower magic
    POISON = "Poison" # lose hp
    DRAIN = "Sap" # lose mp
    FREEZE = "Freeze" # lose a turn

    # Positive
    STRENGTHEN = "Strengthen" # boost attack
    FORTIFY = "Fortify" # boost defense
    HASTEN = "Hasten" # boost speed
    EMPOWER = "Empower" # boost magic
    REGENERATE = "Regenerate" # heal hp
    FOCUS = "Focus" # heal mp
    SLIPPERY = "Slippery" # two consecutive turns

    def __repr__(self):
        return self.value
    def __str__(self):
        return self.value

    def __hash__(self):
        return hash(self)
    '''
    def __eq__(self, other):
        if type(self) == type(other):
            return self == other
        else:
            return False'''

'''    def get_duration(self):
        return self[1]

    def set_duration(self, duration):
        self[1] = duration

    def increment_duration(self):
        self[1] += 1

    def decrement_duration(self):
        self[1] -= 1
        if self[1] < 0:
            self[1] = 0'''

File: java ver/rpg/src/main/java/java_rpg/other/StatusEffect.java
package java_rpg.other;

public enum StatusEffect {
    NORMAL("Normal"),
    // negative
    WEAKEN("Weaken"), // lower attack
    DECAY("Decay"), // lower defense
    DISQUIET("Disquiet"), // lower mp
    SLOW("Slow"), // lower speed
    POISON("Poison"), // lose hp
    SAP("Sap"), // lose mp
    FREEZE("Freeze"), // lose a turn

    // positive
    STRENGTHEN("Strengthen"), // boost attack
    FORTIFY("Fortify"), // boost defense
    EMBOLDEN("Embolden"), // boost magic
    HASTEN("Hasten"), // boost speed
    REGENERATE("Regenerate"), // heal hp
    FOCUS("Focus"), // heal mp
    MELT("Melt"); // 2 consecutive turns

    private final String name;
    /**
     * How long the status effect lasts
     */
    private int duration;

    private StatusEffect(String name) {
        this.name = name;
        this.duration = 0;
    }

    public int getDuration() {
        return this.duration;
    }

    public void setDuration(int duration) {
        this.duration = duration;
        this.duration = this.duration < 0 ? 0 : this.duration;
    }

    public void incrementDuration() {
        this.duration++;
    }

    public void decrementDuration() {
        this.duration--;
        this.duration = this.duration < 0 ? 0 : this.duration;
    }

    public void resetDuration() {
        this.duration = 0;
    }


    @Override
    public String toString() {
        return this.name;
    };

    @Override
    public int hashCode() {
        return this.name.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Stat other) {
            return this.name.equals(other.name);
        } else {
            return false;
        }
    }
}
